File: api/src/main/java/com/wbrawner/twigs/permission/UserPermission.java
package com.wbrawner.twigs.permission;

import com.wbrawner.twigs.budget.Budget;
import com.wbrawner.twigs.user.User;

import javax.persistence.*;

@Entity
public class UserPermission {
    @EmbeddedId
    private UserPermissionKey id;
    @ManyToOne
    @MapsId("budgetId")
    @JoinColumn(nullable = false, name = "budget_id")
    private Budget budget;
    @ManyToOne
    @MapsId("userId")
    @JoinColumn(nullable = false, name = "user_id")
    private User user;
    @JoinColumn(nullable = false)
    @Enumerated(EnumType.STRING)
    private Permission permission;

    public UserPermission() {
        this(null, null, null, null);
    }

    public UserPermission(Budget budget, User user, Permission permission) {
        this(new UserPermissionKey(budget.getId(), user.getId()), budget, user, permission);
    }

    public UserPermission(UserPermissionKey userPermissionKey, Budget budget, User user, Permission permission) {
        this.id = userPermissionKey;
        this.budget = budget;
        this.user = user;
        this.permission = permission;
    }

    public UserPermissionKey getId() {
        return id;
    }

    public Budget getBudget() {
        return budget;
    }

    public User getUser() {
        return user;
    }

    public Permission getPermission() {
        return permission;
    }
}

File: src/main/kotlin/com/wbrawner/budgetserver/permission/UserPermission.kt
package com.wbrawner.budgetserver.permission

import com.wbrawner.budgetserver.budget.Budget
import com.wbrawner.budgetserver.user.User
import com.wbrawner.budgetserver.user.UserResponse
import java.io.Serializable
import javax.persistence.*

@Entity
data class UserPermission(
        @EmbeddedId
        val id: UserPermissionKey? = null,
        @ManyToOne
        @MapsId("budgetId")
        @JoinColumn(nullable = false, name = "budget_id")
        val budget: Budget? = null,
        @ManyToOne
        @MapsId("userId")
        @JoinColumn(nullable = false, name = "user_id")
        val user: User? = null,
        @JoinColumn(nullable = false)
        @Enumerated(EnumType.STRING)
        val permission: Permission? = null
) {
    constructor(budget: Budget, user: User, permission: Permission) : this(UserPermissionKey(budget.id, user.id), budget, user, permission)
}

@Embeddable
data class UserPermissionKey(
        var budgetId: Long? = null,
        var userId: Long? = null
) : Serializable

enum class Permission {
    READ,
    WRITE,
    OWNER
}

data class UserPermissionResponse(
        val user: UserResponse,
        val permission: Permission
) {
    constructor(userPermission: UserPermission) : this(UserResponse(userPermission.user!!), userPermission.permission!!)
}

data class UserPermissionRequest(
        val user: Long,
        val permission: Permission
)
